News Feed

JegorKuz opened the pull request.

JegorKuz committed:
[UPDATE] BJS2-77755
Закончил "Публикация эвента коммента в Kafka".

JegorKuz committed:
[UPDATE] News Feed
Initial commit.
Kafka config.

JegorKuz committed:
Merge branch 'kelpie-stream10-jegor' into kelpie-stream10-jegor-BJS2-77755

JegorKuz committed:
Merge pull request #1981 from CorporationX/kelpie-stream10-jegor-BJS2-77755

[UPDATE] BJS2-77755

JegorKuz committed:
[UPDATE] BJS2-77757
Закончил задачу "Публикация эвента поста в Kafka".

JegorKuz committed:
Merge pull request #1983 from CorporationX/kelpie-stream10-jegor-BJS2-77757

[UPDATE] BJS2-77757

JegorKuz committed:
[UPDATE] BJS2-77757
Закончил задачу "Публикация эвента просмотра поста в Kafka".

JegorKuz committed:
[UPDATE] BJS2-77758
Поправил задачу "Публикация эвента просмотра поста в Kafka".

JegorKuz committed:
Merge pull request #1984 from CorporationX/kelpie-stream10-jegor-BJS2-77758

[UPDATE] BJS2-77758

JegorKuz committed:
[UPDATE] BJS2-77765
Поправил задачу "Публикация эвента лайка в Kafka".
Поправил получаемое дто для KafkaPostViewedEventConsumer

JegorKuz committed:
Merge pull request #1985 from CorporationX/kelpie-stream10-jegor-BJS2-77765

[UPDATE] BJS2-77765

JegorKuz committed:
[UPDATE] BJS2-77764
Добавлена подготовительная логика для получения ид всех пользователей.

JegorKuz committed:
[UPDATE] BJS2-77764
Добавлены Consumer/Producer для эвента feed-heat.
Закончен цикл начала заполнения/прогревания хэша.

JegorKuz committed:
[UPDATE] BJS2-77764
Исправил помарки.
Выделил в ConsumerGroup: feed-heat-group, теперь реплики сервиса будут разбирать батчи без дублирования.

JegorKuz committed:
[UPDATE] BJS2-77764
Исправил помарки
Настроил Redis для кэша.

JegorKuz committed:
[UPDATE] BJS2-77764
Внедрил убогое решение для разогрева кэша.
Надо бы улучшить, а как?)

JegorKuz committed:
[UPDATE] BJS2-77764
Закончил разогрев кэша, можно ещё разбить подписчиков на батчи.
Если у 10 человек 1 миллион подписчиков, при запросе тысячи пользователей, вернется слишком много подписчиков, каждому пихать через фор луп долго, можно разбить особо большие списки подписчиков на сообщения для кафки.

JegorKuz committed:
[UPDATE] BJS2-77761
Пост сервис отвечает только за отправку сообщения о посте.
Фид сервис отвечает за сбор подписчиков и обновление их фида.

JegorKuz committed:
Merge pull request #2003 from CorporationX/kelpie-stream10-jegor-77761

Kelpie stream10 jegor 77761

JegorKuz committed:
[UPDATE] BJS2-77760
Сохранение автора коммента в Redis после публикации.

JegorKuz committed:
Merge pull request #2004 from CorporationX/kelpie-stream10-jegor-BJS2-77760

[UPDATE] BJS2-77760
Исправить!
Кэш сохраняет пользователя как - String:long. поменять на String:UserDto

JegorKuz commented:
[UPDATE] BJS2-77760
Исправить!
Кэш сохраняет пользователя как - String:long. поменять на String:UserDto

JegorKuz committed:
[UPDATE] BJS2-77760
Сохранение автора поста в Redis после публикации.

JegorKuz committed:
Merge pull request #2005 from CorporationX/kelpie-stream10-jegor-BJS2-77767

[UPDATE] BJS2-77767
Сохранение автора поста в Redis после публикации.

JegorKuz committed:
[UPDATE] BJS2-77762
Слушатель Kafka-эвентов для комментов.

JegorKuz committed:
Merge pull request #2008 from CorporationX/kelpie-stream10-jegor-BJS2-77762

[UPDATE] BJS2-77762

JegorKuz committed:
[UPDATE] BJS2-77756
Слушатель Kafka-эвентов для просмотров поста.
Создал отдельный ключ для хранения счётчика просмотров поста.

JegorKuz committed:
Merge pull request #2009 from CorporationX/kelpie-stream10-jegor-BJS2-77756

[UPDATE] BJS2-77756 Слушатель Kafka-эвентов для просмотров поста. Создал отдельный ключ для хранения счётчика просмотров поста.

JegorKuz committed:
[UPDATE] BJS2-77756
Слушатель Kafka-эвентов для постов.
Добавил атомарную обработку нового поста для подписчиков.
Надо добавить ack после того как все подписчики получили пост, сделать асинхронно через CompletableFuture и дождаться через Join().

JegorKuz committed:
Merge pull request #2010 from CorporationX/kelpie-stream10-jegor-BJS2-77759

[UPDATE] BJS2-77756 Слушатель Kafka-эвентов для постов. Добавил атомарную обработку нового поста для подписчиков. Надо добавить ack после того как все подписчики получили пост, сделать асинхронно через CompletableFuture и дождаться через Join().

JegorKuz committed:
[UPDATE] BJS2-77763
Слушатель Kafka-эвентов для лайков
Добавил атомарную обработку нового лайка для поста.

JegorKuz committed:
Merge pull request #2011 from CorporationX/kelpie-stream10-jegor-BJS2-77763

[UPDATE] BJS2-77763 Слушатель Kafka-эвентов для лайков Добавил атомарную обработку нового лайка для поста.

miked4king commented on src/main/java/faang/school/postservice/cache/NewsFeedCache.java:
```diff
@@ -0,0 +1,245 @@
+    private static final String FEED_CACHE_KEY_PREFIX = "feed_cache:";
+    private static final String POST_COMMENTS_KEY_PREFIX = "comments_cache:";
+    private static final String POST_VIEWS_KEY_PREFIX = "post_views:";
+    private static final String POST_LIKES_KEY_PREFIX = "post_likes:";
```
как то много кэшей, вроде 3 по условию должно быть

JegorKuz replied:
Согласен, переделал.

miked4king commented on src/main/java/faang/school/postservice/cache/NewsFeedCache.java:
```diff
@@ -0,0 +1,245 @@
+    private final RedisTemplate<String, Object> redisNewsFeedTemplate;
+    private final ObjectMapper objectMapper;
+
+    @Value("${spring.data.redis-news-feed.size}")
+    private int feedSizeLimit;
+    @Value("${spring.data.redis-news-feed.user-ttl}")
+    private int userTTL;
+    @Value("${spring.data.redis-news-feed.comments-limit}")
+    private int commentsLimit;
```
в отдельный класс с configurationProperties

miked4king commented on src/main/java/faang/school/postservice/cache/NewsFeedCache.java:
```diff
@@ -0,0 +1,245 @@
+        // Скрипт для комментариев к посту
+        addCommentAndTrimScript = new DefaultRedisScript<>();
+        addCommentAndTrimScript.setScriptText("""
+                     local key = KEYS[1]
+                     local limit = tonumber(ARGV[1])
+                     local score = tonumber(ARGV[2])
+                     local member = ARGV[3]
+                     -- ZADD идемпотентен.
+                     redis.call('ZADD', key, score, member)
+                     -- ZREMRANGEBYRANK 0, -limit-1 удаляет все элементы, кроме последних 'limit'.
+                     redis.call('ZREMRANGEBYRANK', key, 0, -limit-1)
+                     return redis.call('ZCARD', key)
```
сходу сложно понять что за скрипты и для чего предназначены, как правило в коде не хранят.
можешь отдельные файлы в проекте сделать, описать для чего предназначены

miked4king commented on src/main/java/faang/school/postservice/cache/NewsFeedCache.java:
```diff
@@ -0,0 +1,245 @@
+    public void putUser(Long user) {
+        String key = USER_CACHE_KEY_PREFIX + user;
+        redisNewsFeedTemplate.opsForValue().set(key, user, Duration.ofMinutes(5));
+//        redisNewsFeedTemplate.opsForValue().set(key, user, Duration.ofDays(userTTL)); // не забыть раскоментить после тестов.
```
комменты убрать

miked4king commented on src/main/java/faang/school/postservice/config/kafka/KafkaTopicConfig.java:
```diff
@@ -0,0 +1,69 @@
+    @Bean
+    public KafkaAdmin kafkaAdmin() {
+        Map<String, Object> configs = new HashMap<>();
+        configs.put(org.apache.kafka.clients.admin.AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
```
org.apache.kafka.clients.admin. ?

miked4king commented on src/main/java/faang/school/postservice/config/kafka/KafkaTopicConfig.java:
```diff
@@ -0,0 +1,69 @@
+
+    @Bean
+    public NewTopic feedHeatEventTopic() {
+        return TopicBuilder.name("feed_heat_event")
```
для большей гибкости вынести в конфигурацию названия топиков

miked4king commented on src/main/java/faang/school/postservice/config/kafka/KafkaTopicConfig.java:
```diff
@@ -0,0 +1,69 @@
+@Component
+public class KafkaTopicConfig {
+
+    @Value("${spring.kafka.bootstrap-servers}")
+    private String bootstrapServers;
+    @Value("${spring.kafka.variables.partitions}")
+    private int partitions;
+    @Value("${spring.kafka.variables.replicas}")
+    private int replicas;
```
@ConfigurationProperties

miked4king commented on src/main/java/faang/school/postservice/config/redis/feed/RedisFeedConfiguration.java:
```diff
@@ -0,0 +1,52 @@
+        RedisTemplate<String, Object> template = new RedisTemplate<>();
+        template.setConnectionFactory(redisNewsFeedConnectionFactory);
+
+        // Создаем и настраиваем ObjectMapper
+        ObjectMapper objectMapper = new ObjectMapper();
+
+        // Регистрируем модуль для поддержки Java 8 Time API
+        objectMapper.registerModule(new JavaTimeModule());
+
+        // Отключаем сериализацию полей-дат как timestamp-чисел для лучшей читаемости в Redis
+        objectMapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
+
+        // Создаем сериализатор Jackson с настроенным ObjectMapper
+        GenericJackson2JsonRedisSerializer jackson2JsonRedisSerializer = new GenericJackson2JsonRedisSerializer(objectMapper);
+
+        // Устанавливаем настроенный сериализатор для значений
```
gpt?)
убираем комменты

JegorKuz replied:
gemini :D

miked4king commented on src/main/java/faang/school/postservice/config/redis/feed/RedisFeedProperties.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.stereotype.Component;
+
+@Component
```
@ConfigurationPropertiesScan на мэйн класс и не придется @Component накидывать на конф классы

miked4king commented on src/main/java/faang/school/postservice/config/redis/threads/RedisThreadPoolConfig.java:
```diff
@@ -0,0 +1,23 @@
+@RequiredArgsConstructor
+public class RedisThreadPoolConfig {
+
+    private final RedisThreadPoolProperties redisThreadPoolProperties;
```
можем инжектить в параметры redisTaskExecutor

miked4king commented on src/main/java/faang/school/postservice/controller/feed/FeedController.java:
```diff
@@ -0,0 +1,27 @@
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+
+@RequestMapping("/api/v1/feed")
```
/api/v1 - конфигурация

miked4king commented on src/main/java/faang/school/postservice/controller/feed/FeedController.java:
```diff
@@ -0,0 +1,27 @@
+
+@RequestMapping("/api/v1/feed")
+@RequiredArgsConstructor
+@Controller
```
почему просто Controller ?

miked4king commented on src/main/java/faang/school/postservice/controller/feed/FeedController.java:
```diff
@@ -0,0 +1,27 @@
+
+    @GetMapping
+    public void getFeed(NewsFeedRequestDto dto) {
+
```
?

miked4king commented on src/main/java/faang/school/postservice/kafka/consumer/KafkaCommentEventConsumer.java:
```diff
@@ -0,0 +1,24 @@
+    @Value("${spring.kafka.topics.comment-event}")
+    public final String topic;
+
+    @KafkaListener(topics = "#{__listener.topic}", groupId = "my-group")
```
почему такой синтакс #{__listener.topic} ? откуда тянем ?
my-group - не очень для названия группы)

JegorKuz replied:
#{__listener.topic}
'#'  - специфический синтаксис кафки.
listener - обращение к самому себе по сути, как this
topic - название параметра в listener

miked4king commented on src/main/java/faang/school/postservice/kafka/consumer/KafkaCommentEventConsumer.java:
```diff
@@ -0,0 +1,24 @@
+
+    @KafkaListener(topics = "#{__listener.topic}", groupId = "my-group")
+    public void listen(KafkaCommentEventDto dto) {
+        System.out.println("Received message: " + dto);
```
?

mysosnovskaya replied:
Не исправлено. System.out.println мы не используем в проектах. Только логгер, если нужно

miked4king commented on src/main/java/faang/school/postservice/kafka/consumer/KafkaCommentEventConsumer.java:
```diff
@@ -0,0 +1,24 @@
+    @KafkaListener(topics = "#{__listener.topic}", groupId = "my-group")
+    public void listen(KafkaCommentEventDto dto) {
+        System.out.println("Received message: " + dto);
+        feedService.putUserIntoCache(dto.authorId());
+        feedService.updateCommentInCache(dto);
```
ack тут не нужен?

miked4king commented on src/main/java/faang/school/postservice/kafka/consumer/KafkaFeedHeatEventConsumer.java:
```diff
@@ -0,0 +1,36 @@
+        try {
+            feedService.fillCacheForUsers(dto.users());
+
+            // Если обработка прошла успешно, отправляем подтверждение
```
убрать комменты

miked4king commented on src/main/java/faang/school/postservice/kafka/consumer/KafkaLikeEventConsumer.java:
```diff
@@ -0,0 +1,22 @@
+
+    @KafkaListener(topics = "#{__listener.topic}", groupId = "my-group")
+    public void listen(KafkaLikeEventDto dto) {
+        System.out.println("Received message: " + dto);
```
?

miked4king commented on src/main/java/faang/school/postservice/kafka/consumer/KafkaPostEventConsumer.java:
```diff
@@ -0,0 +1,24 @@
+    @Value("${spring.kafka.topics.post-event}")
+    public final String topic;
+
+    @KafkaListener(topics = "#{__listener.topic}", groupId = "my-group")
+    public void listen(KafkaPostEventDto dto) {
+        System.out.println("Received message: " + dto);
```
?

mysosnovskaya replied:
Аналогично про my-group и System.out.println

miked4king commented on src/main/java/faang/school/postservice/kafka/consumer/KafkaPostViewedEventConsumer.java:
```diff
@@ -0,0 +1,23 @@
+
+    @KafkaListener(topics = "#{__listener.topic}", groupId = "my-group")
+    public void listen(KafkaPostViewedEventDto dto) {
+        System.out.println("Received message: " + dto);
```
?

miked4king commented on src/main/java/faang/school/postservice/kafka/producer/KafkaCommentEventProducer.java:
```diff
@@ -0,0 +1,20 @@
+
+@Component
+@RequiredArgsConstructor
+public class KafkaCommentEventProducer {
```
интерфейс?

miked4king commented on src/main/java/faang/school/postservice/service/feed/FeedServiceImpl.java:
```diff
@@ -0,0 +1,99 @@
+    @Value("${news-feed.heater.batch-size}")
+    private int batchSize;
+
+    // PostConstruct?
```
насчет postconstruct не уверен, если будет туча пользователей очень долго будет сервис подниматься

miked4king commented on src/main/java/faang/school/postservice/service/feed/FeedServiceImpl.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    public void fillCacheForUsers(List<UserFeedHeatDto> users) {
+        // Заполнение будет происходить через пользователя и его подписчиков.
```
Егор, много вопросов)
твоя задача самостоятельно пропработать решение и реализовать его, если что посмотри лекции

JegorKuz replied:
все коменты в коде для себя оставлены, что-бы не забыть про это

JegorKuz committed:
[UPDATE] Получение news-feed
NewsFeedHeat:
1) Получить пользователей батчами.
2) Батч пользователей использовать что-бы вытягивать подписчиков каждого пользователя так же батчами и отправить в кафку для распределения между инстансами сервисов.
3) Если пользователь обладает подписчиками то добавляем его в кэш.
4) Для батча подписчиков закинуть посты пользователя в фид.
Comment/Like/View Event:
1) Послать\поймать дто эвента в кафку.
2) Добавить в кэш. (Comment)
3) Обновить данные поста.
Post Event:
1) Послать\поймать дто эвента в кафку.
2) Добавить в кэш пост и автора.
3) Найти всех подписчиков автора.
4) Добавить пост в фид подписчиков.
Get Feed for user:
1) По ид пользователя вытащить 20 постов исходя из переданного параметра.
2) До собрать посты из кэша. (комменты)
3) Вернуть пользователю посты

Добавить:
1) Acknowledgement.
2) Обработчик ошибок.
3) Retryable
4) Можно ли где-то в кэше добавить асинхронность?
5) Доставать батч постов по их ид?
6) Когда сущность удаляется по TTL надо будет по чистить другие части рэдис где эта сущность использовалась?
Если исчез пост, зачем хранить коменты к нему?
Отлавливать через рэдис и чистить кэш.

JegorKuz committed:
Merge pull request #2018 from CorporationX/kelpie-stream10-jegor-BJS2-77766

[UPDATE] Получение news-feed
NewsFeedHeat:
1) Получить пользователей батчами.
2) Батч пользователей использовать что-бы вытягивать подписчиков каждого пользователя так же батчами и отправить в кафку для распределения между инстансами сервисов.
3) Если пользователь обладает подписчиками то добавляем его в кэш.
4) Для батча подписчиков закинуть посты пользователя в фид.
Comment/Like/View Event:
1) Послать\поймать дто эвента в кафку.
2) Добавить в кэш. (Comment)
3) Обновить данные поста.
Post Event:
1) Послать\поймать дто эвента в кафку.
2) Добавить в кэш пост и автора.
3) Найти всех подписчиков автора.
4) Добавить пост в фид подписчиков.
Get Feed for user:
1) По ид пользователя вытащить 20 постов исходя из переданного параметра.
2) До собрать посты из кэша. (комменты)
3) Вернуть пользователю посты

Добавить:
1) Acknowledgement.
2) Обработчик ошибок.
3) Retryable
4) Можно ли где-то в кэше добавить асинхронность?
5) Доставать батч постов по их ид?
6) Когда сущность удаляется по TTL надо будет по чистить другие части рэдис где эта сущность использовалась?
Если исчез пост, зачем хранить коменты к нему?
Отлавливать через рэдис и чистить кэш.

JegorKuz committed:
[UPDATE] News Feed
1) Создал новый запрос в UserService что-бы получать пользователя с его фотографией.
2) Оптимизировал получение батча постов в кэше.
3) Добавил новый lua-script и логику для добавления батча постов.
4) Теперь вместо ид в редис добавляется дто пользователя.
5) Добавил логику на случай если поста нет в кэше.

СДЕЛАТЬ!
Почистить от лишних методов, переменных, комментариев.

JegorKuz committed:
[UPDATE] News Feed
в FeedService:
1) Убрал дублирование кода для получения подписчиков батчами.
2) Добавил фолбэк на случай падения UserService при запросе на получения пользователей.
3) Добавил "красивую" обработку try-catch.
4) Добавил коментарии к фиду, теперь сборка фида завершена.

в NewsFeedCache
1) Убрал неиспользуемые\заменённые методы.
2) Добавил логику получения коментариев.

Почистил от лишних методов и комментариев.

Вынести луа скрипты отдельно. Протестить юзкейсы решение.

JegorKuz committed:
[UPDATE] News Feed
NewsFeedCache:
1) Вынес lua-script-ы из кэша в ресурсы.
2) Вынес константы в конфиг.
KafkaTopicConfig - Вынес константы в properties
RedisFeedConfiguration - Почистил коменты
По правил помарки связанные с ревью.

Протестить юзкейсы решения.

JegorKuz committed:
[UPDATE] News Feed
Протестил юзкейсы решения:
1) Заполнения фида. (/heat) [DONE]
1.1) Добавлен дополнительный RedisTemplate<String, String> для lua-запросов требующих String[] в качестве args.
1.2) Изменён запрос для получения батча постов пользователя.
1.2.1) Для маппинга получаемых постов из бд, нужен отдельный конструктов из-за поля "comments".
1.3) Поправил запросы в UserService.
1.4) Что-бы избежать дублирования запросов на сохранение пользователей в редис, сохранение было вынесено в метод через который каждый из пользователей проходит единожды.
1.5) При запросе на разогрев кэша поток содержащий x-user-id в header, передаёт батчи пользователей в кафку, поток меняется при получении сообщения consumer-ом. Что-бы запросы в UserService могил выполнится успешно перед началом while - устанавливается user-id для userContext.
1.6) Для установления батчу постов ttl теперь конвертируется в секунды и устанавливается отдельным запросом в lua-script.
2) Получение фида:
2.1) с postId
2.2) первые 20
3) Эвенты
3.1) коммент
3.2) публикация поста
3.3) лайк поста
3.4) просмотр поста

Добавить запросы в бд в случае если чего-то не существует в кэше.

JegorKuz committed:
[UPDATE] News Feed
Протестил юзкейсы решения:
2) Получение фида: [DONE]
2.1) При сохранении поста в фид, его ид сохраняется как стринга, при получении обратно, стринга автоматически может сконвертироваться в int вместо long. Добавлена доп проверка.
2.2) Добавил новый запрос для нахождения комментариев для поста.
2.3) с postId [DONE]
2.3.1) Сделал - что-бы сначала находился пост с нужным ид, а потом уже бралось 20 следующих постов.
2.4) первые 20 [DONE]
3) Эвенты
3.1) коммент
3.2) публикация поста
3.3) лайк поста
3.4) просмотр поста

Добавить запросы в бд в случае если чего-то не существует в кэше.

JegorKuz committed:
[UPDATE] News Feed
Протестил юзкейсы решения:
3) Эвенты [DONE]
Теперь версионирование происходит внутри луа скрипта что-бы избежать race-condition.
3.1) коммент
3.2) публикация поста
3.2.1) Пост теперь сериализуется до попытки положить его в кэш, тк когда он достаётся используется именноподход со стрингой, значит и класть в кэш надо так же.
3.3) лайк поста
3.4) просмотр поста
3.4.1) Исправил проблему при попытке увеличить счётчик просмотров в луа скрипте, из-за viewCount(null) недопустимое значение выдавало ошибку. Теперь значения присваиваются перед добавлением любого поста в кэш.

Добавить запросы в бд в случае если чего-то не существует в кэше.

JegorKuz committed:
[UPDATE] News Feed
Исправил тест.

mysosnovskaya commented on src/main/java/faang/school/postservice/kafka/consumer/KafkaCommentEventConsumer.java:
```diff
@@ -0,0 +1,24 @@
+    @Value("${spring.kafka.topics.comment-event}")
+    public final String topic;
+
+    @KafkaListener(topics = "#{__listener.topic}", groupId = "my-group")
```
groupId  тоже надо из application.yaml брать. Ну и как Миша писал - my-group плохое название, нужно специфичное

JegorKuz replied:
сделал

mysosnovskaya commented on src/main/java/faang/school/postservice/kafka/consumer/KafkaCommentEventConsumer.java:
```diff
@@ -0,0 +1,24 @@
+    public void listen(KafkaCommentEventDto dto) {
+        System.out.println("Received message: " + dto);
+        feedService.putCommentInCache(dto);
+        feedService.updatePost(dto.postId(), topic);
```
Про ack не увидела ответа на Мишин коммент.

JegorKuz replied:
Там было подготовлено что-бы возвращать boolean для проверки как прошло сохранение, надо было логику подправить под это дело. Теперь сделал.

mysosnovskaya commented on src/main/java/faang/school/postservice/kafka/KafkaEventProducer.java:
```diff
@@ -0,0 +1,5 @@
+package faang.school.postservice.kafka;
+
+public interface KafkaEventProducer<T> {
```
А может быть даже абстрактный класс лучше подошел бы?) По сути все реализации одинаковые получились

mysosnovskaya reviewed (commented):
На сертификат засчитываю, конечно) Но кое-какие доработки ещё требуются

JegorKuz commented:
Если честно, я хотел бы понять, правильно ли вообще моя логика определена?
Есть ли уязвимые места?
Например для меня до сих пор загадка:
При разогреве кэша, запускается do/while который батчами подтягивает из сервиса userId и кладёт в кафку, но что если на 50% сервис с userId упал или сам сервис фида упал, кэш то уже заполнен на половину, как мне отслеживать где я остановился? Или тут не так важно и если он пойдёт крутить по второму разу и снова подтянет всех, то дубликатов то всё равно не будет так что можно не нервничать. Однако если это Meta где миллионы пользователей то крутить по новой может быть долго, тогда что делать? У меня единственный вариант был - это сохранять в Postgres ид батча, с какого по какой ID мы выгрузили. (10.000-20.000) и так мы понимаем где остановились и продолжаем только с этого места, но тогда мы тратим доп время на Update для сущности в бд, что слегка тормозит процесс, или уж лучше так слегка замедлить но за-то сэкономить время в случае падения сервиса?

JegorKuz committed:
[UPDATE] News Feed
Поправил ошибки по коментариям.
